### PR TITLE
dns: Review for DNS implementation

### PR DESCRIPTION
**Pull request by DipSwitch**
Can someone please review my DNS implementation?

Re-Write to use `conn` API depends on #3921 
Re-Writing to `conn` loses the capability of zero-copy since I now could allocate the packet buffer at maximum size and reallocate it to the size that it actually is after filling it.

### CODE REVIEW

File: sys/include/net/gnrc/udp.h [hunk 1]
```diff
@@ -40,7 +40,7 @@ extern "C" {
 #endif
 
 /**
- * @brief   Priority of the pktdump thread
+ * @brief   Priority of the UDP thread
```
Comment (cgundogan) [inline sys/include/net/gnrc/udp.h]: unrelated. Move this into it's own PR (:

File: sys/net/gnrc/application_layer/dns/Makefile [hunk 2]
```diff
@@ -0,0 +1,3 @@
+MODULE = gnrc_dns
```
Comment (cgundogan) [inline sys/net/gnrc/application_layer/dns/Makefile]: you probably don't need this line, could you test without it?

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+/* For PRIu16 etc. */
+#include <inttypes.h>
+#endif
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
```
Comment (cgundogan) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: UDP?

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+static void *_event_loop(void *args);
+
+kernel_pid_t _dns_pid = KERNEL_PID_UNDEF;
+static gnrc_dns_request_t *_dns_requests[GNRC_DNS_MAX_SIMULTANEOUS_REQUESTS];
+static gnrc_dns_servers_t _dns; /**< DNS servers object (TODO currently only one, because we only have one interface) */
+static uint16_t _request_id;
+
+static void _gnrc_dns_delete_request(gnrc_dns_request_t *req) {
+    /* free the request from the running DNS requests */
+    for (int idx = 0; idx < sizeof(_dns_requests[0]); ++idx) {
```
Comment (cgundogan) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: why not just check for `idx < GNRC_DNS_MAX_SIMULTANEOUS_REQUESTS` ? same below

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+    /* check if thread is already running */
+    if (_dns_pid == KERNEL_PID_UNDEF) {
+        /* start DNS thread */
+        _dns_pid = thread_create(_stack, sizeof(_stack), GNRC_DNS_PRIO,
+        CREATE_STACKTEST, _event_loop, NULL, "dns");
+    }
+    return _dns_pid;
+}
+
+int gnrc_dns_print(void) {
```
Comment (cgundogan) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: Is it possible to move this into `DEVELHELP`?

File: sys/net/gnrc/netreg/gnrc_netreg.c [hunk 6]
```diff
@@ -119,10 +119,6 @@ int gnrc_netreg_calc_csum(gnrc_pktsnip_t *hdr, gnrc_pktsnip_t *pseudo_hdr)
         case GNRC_NETTYPE_ICMPV6:
             return gnrc_icmpv6_calc_csum(hdr, pseudo_hdr);
 #endif
-#ifdef MODULE_GNRC_TCP
```
Comment (cgundogan) [inline sys/net/gnrc/netreg/gnrc_netreg.c]: why do you remove this here?

File: sys/include/net/gnrc/dns/internals.h [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    uint16_t request_id;
+    dns_query_type query_type;
+    gnrc_dns_response_t *response;
+} gnrc_dns_request_t;
+
+/**
+ * @brief   DNS server advertise information
+ */
+typedef struct {
+    mutex_t mutex;                                  /**< DNS mutex */
```
Comment (cgundogan) [inline sys/include/net/gnrc/dns/internals.h]: what do you need this mutex for? You only have one thread, or do you intend to share your resource with other layers?

File: sys/net/gnrc/application_layer/dns/Makefile (hunk 2 above)
Comment (authmillenon) [inline sys/net/gnrc/application_layer/dns/Makefile]: He needs it because the directory name is `dns`, not `gnrc_dns` ;-)

File: sys/include/net/gnrc/dns/internals.h (hunk 7 above)
Comment (DipSwitch) [inline sys/include/net/gnrc/dns/internals.h]: The DNS servers are passed from NDP asynchronous.

File: sys/include/net/gnrc/udp.h (hunk 1 above)
Comment (DipSwitch) [inline sys/include/net/gnrc/udp.h]: Will do tomorrow. :)

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c (hunk 5 above)
Comment (DipSwitch) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: Don't know. This function is used by the shell also.

File: sys/net/gnrc/netreg/gnrc_netreg.c (hunk 6 above)
Comment (DipSwitch) [inline sys/net/gnrc/netreg/gnrc_netreg.c]: This is not yet in master if i'm correct?

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c (hunk 5 above)
Comment (authmillenon) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: You could put it to its own *.c file, then the linker will only include it, when it is used, regardless of optimization settings.

Comment (cgundogan) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: +1

Comment (DipSwitch) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: If you compile with `-ffunction-sections` and link with `-Xlinker --gc-sections` it won't be included in the build unless it's used. Also when optimization is disabled.

	0) dipswitch@dipswitch-K55VD:~/RIOT/examples/gnrc_networking $ grep gnrc_dns_print bin/ax-avior/gnrc_networking.map 
	 .text.gnrc_dns_print
		        0x00000000080016a0                gnrc_dns_print
	gnrc_dns_print                                    /home/dipswitch/RIOT/examples/gnrc_networking/bin/ax-avior/gnrc_dns.a(gnrc_dns.o)

After removing the call from `sc_gnrc_dns.c`

	0) dipswitch@dipswitch-K55VD:~/RIOT/examples/gnrc_networking $ grep gnrc_dns_print bin/ax-avior/gnrc_networking.map 
	 .text.gnrc_dns_print
	gnrc_dns_print                                    /home/dipswitch/RIOT/examples/gnrc_networking/bin/ax-avior/gnrc_dns.a(gnrc_dns.o)

I can create two calls like `bool gnrc_dns_claim_servers(int *count, ipv6_addr_t **addr, *ttl)` and `void gnrc_dns_release_servers()`
Or more my favorite since it's better usable in the rest of the system `bool gnrc_dns_get_server(int idx, gnrc_server_info_t *s)` where you get one server out of the collection and then move the print to the shell file. And you can't forget to release mutex :)

Comment (DipSwitch) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: Maybe the new implementation is beter :)

**Comment by DipSwitch** [comment]
Implementation is finished if you ask me :) maybe sending the RTR_ADV needs some attention to get the DNS servers and send them in his RTR_ADV.

**Comment by authmillenon** [comment]
Great! But please know, that we are currently in feature freeze for the release, so a proper review might take a while.

**Comment by DipSwitch** [comment]
Np, enough other stuff to make ^_^

**Comment by authmillenon** [comment]
@DipSwitch may it be possible to use the conn interface for application protocols? This way other future network stacks would also profit from your implementation.

**Comment by DipSwitch** [comment]
For DNS I also need to be able to receive a timeout callback from the timer if the DNS "leases" expire. This can also be done from an `xtimer` callback (scheduled from the idle thread for example) which runs from another thread since I a mutex to protect my global variables.
Or I can change it's behaviour to check if the DNS server has expired before I make a new DNS request.

**Comment by gebart** [comment]
needs rebase

**Comment by DipSwitch** [comment]
How about this one? I don't have the time to rewrite to `conn` at the moment. I can however later rewrite it if we add it this release using the low level API.

**Comment by OlegHahm** [comment]
IMO it's fine not to use conn for now. We may consider changing/replacing conn with something simpler anyway IMO. (Just a personal opinion so far.)

**Comment by DipSwitch** [comment]
In that case no need to wait for another PR :) Need to squash?

**Comment by OlegHahm** [comment]
Yes, I think you should squash.

**Comment by DipSwitch** [comment]
Rebased and squashed

**Comment by emmanuelsearch** [comment]
@DipSwitch can you tell me how to test this best? Would like to try it. Would be great if this could make the next release.

**Comment by DipSwitch** [comment]
I Tested it with the following `radvd.conf` over 6LowPAN

	root@raspberrypi:~# cat /etc/radvd.conf
	interface eth0
	{
		AdvSendAdvert on;
		#AdvManagedFlag on;
		#AdvOtherConfigFlag on;

		route fdcb:61::/64 {
		};
	};

	interface lowpan0
	{
		AdvSendAdvert on;
		AdvManagedFlag on;
		AdvOtherConfigFlag on;

		prefix fdcb:61::/64 {
			AdvOnLink off;
			AdvAutonomous on;
			#AdvRouterAddr on; # for IPv6 Mobile
		};
	#    RDNSS fdcb:172:31::1:254 {
	#        AdvRDNSSLifetime 600;
	#    };
		RDNSS fdcb:61::1 {
			AdvRDNSSLifetime 600;
		};
	};

And with the following `radvd.conf` on native with a tap device.

	0)es_user@esubuntu:~ $ cat radvd.conf 
	interface eth0
	{
		AdvSendAdvert on;
		#AdvManagedFlag on;
		#AdvOtherConfigFlag on;

		route fdcb:61::/64 {
		};
	};

	interface tapbr0
	{
		AdvSendAdvert on;
		AdvManagedFlag on;
		AdvOtherConfigFlag on;

		prefix fdcb:61::/64 {
			AdvOnLink on;
		AdvAutonomous on;
		#AdvRouterAddr on; # for IPv6 Mobile
		};
		RDNSS fdcb:172:31::1:254 {
			AdvRDNSSLifetime 600;
		};
	};

**Comment by DipSwitch** [comment]
You can install `bind9` on the raspberry PI and configure it for forwarding to a IPv4 DNS server.

	root@raspberrypi:~# cat /etc/bind/named.conf.options
	options {
			directory "/var/cache/bind";

			forwarders {
					8.8.8.8;
			};

			dnssec-validation auto;

			auth-nxdomain no;    # conform to RFC1035
			listen-on-v6 { any; };
	};

**Comment by DipSwitch** [comment]
@emmanuelsearch Is the 'explanation' clear? Or need some more help?

One thing I forgot is that after the DNS is properly configured on the node you can test a DNS query via the console: `dns res www.google.com` for the AAAA record and `dns res www.google.com A` for the A record for example.

If you have an DNS server working on IPv6 behind the boarder router (normal network) you could also just set it's IP via the RIOT shell `dns add fd**:**:**::*`

**Comment by emmanuelsearch** [comment]
Sorry, got swamped with other stuff. Will try to test next week!

**Comment by OlegHahm** [comment]
Needs a rebase.

**Comment by zhuoshuguo** [comment]
@DipSwitch Ping. A kind reminder for rebase.

**Comment by DipSwitch** [comment]
Sorry swamped with work, I'll have some time this weekend to check all the responses and update the codes.

**Comment by DipSwitch** [comment]
Merged with latest

File: sys/include/net/gnrc/dns.h [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+#define GNRC_DNS_PRIO                        (THREAD_PRIORITY_MAIN - 2)
+#endif
+
+/**
+ * @def GNRC_DNS_STACK_SIZE
+ *
+ * @brief   Default stack size to use for the DNS thread
+ */
+#ifndef GNRC_DNS_STACK_SIZE
+#define GNRC_DNS_STACK_SIZE                  (2048)
```
Comment (OlegHahm) [inline sys/include/net/gnrc/dns.h]: Wouldn't it be preferable to use a value based on a platform dependent value here like `THREAD_STACKSIZE_DEFAULT + X` as we do for the other GNRC stack sizes?

File: sys/include/net/gnrc/dns/internal.h [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+/**
+ * @brief   Compile time byte-order swapping for short constants
+ */
+#define CT_HTONS(x)    ((uint16_t)x)
+
+/**
+ * @brief   Compile time byte-order swapping for integers constants
+ */
+#define CT_HTONL(x)    ((uint32_t)x)
+#endif
```
Comment (OlegHahm) [inline sys/include/net/gnrc/dns/internal.h]: Shouldn't these defines be removed here and make the PR dependent on #4812 instead?

File: sys/include/net/gnrc/dns/internal.h [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+/**
+ * @brief   DNS types
+ * @{
+ */
+typedef uint16_t dns_query_type;
+
+#define DNS_A                   CT_HTONS(0x0001)    /**< IPv4 */
+#define DNS_CNAME               CT_HTONS(0x0005)    /**< CNAME */
+#define DNS_TXT                 CT_HTONS(0x0010)    /**< Text */
+#define DNS_AAAA                CT_HTONS(0x001C)    /**< IPv6 */
```
Comment (OlegHahm) [inline sys/include/net/gnrc/dns/internal.h]: Why having the `GNRC_DNS_TYPE_*` enums additionally?

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+            return true;
+        }
+    }
+
+    return false;
+}
+
+gnrc_dns_servers_t *_gnrc_dns_get_context(void)
+{
+    return &_dns;
```
Comment (OlegHahm) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: Just make `_dns` a global variable?

File: sys/include/net/gnrc/dns.h (hunk 8 above)
Comment (DipSwitch) [inline sys/include/net/gnrc/dns.h]: Addressed

File: sys/include/net/gnrc/dns/internal.h (hunk 10 above)
Comment (DipSwitch) [inline sys/include/net/gnrc/dns/internal.h]: Because these are the DNS query types which must be send to the server. The `gnrc_dns_type_t` enum is exported outside of the module and I didn't want to bloat the source with global `CT_HTONS` etc. Also since this is an enum the type doesn't need to be `uint16_t`

File: sys/net/gnrc/application_layer/dns/gnrc_dns.c (hunk 11 above)
Comment (DipSwitch) [inline sys/net/gnrc/application_layer/dns/gnrc_dns.c]: Jup :)

**Comment by DipSwitch** [comment]
I could modify the `dist/tools/tabsetup` script to also assign a site-local address to the net addapters 
and start a `radvd` instance with a custom configuration file which assigns a DNS server via `RDNSS` to a native node.
Maybe even start a local `bind9` server which forwards the packets to the google public DNS.

If (and it doesn't) break NDP we could just merge it and let users test it if we add an howto in the readme section?

**Comment by DipSwitch** [comment]
Rebase and squashed

**Comment by kYc0o** [comment]
The conn API still WIP so this one would be postponed.

**Comment by cgundogan** [comment]
@DipSwitch does this PR still waits for another PR? #3921  was closed without merging

**Comment by DipSwitch** [comment]
It worked the last time I tested it. And rewriting later isn't a problem :)

**Comment by miri64** [comment]
FYI the sock API introduced in #5533 (rename will happen today) supports timeouts natively.

**Comment by miri64** [comment]
@DipSwitch ping?

**Comment by miri64** [comment]
GNRC port for `sock` is available at #5772. Would it be a problem porting it to that one?

**Comment by miri64** [comment]
I heard @kaspar030 is working on a DNS port based on `sock`. Is that correct?

**Comment by DipSwitch** [comment]
Sorry fall of the radar for a while... I can rebase this weekend, tested it a few weeks ago and still works like a charm :)

**Comment by miri64** [comment]
Postponed due to feature freeze.

**Comment by miri64** [comment]
@kaspar030 @DipSwitch ping?

**Comment by kaspar030** [comment]
> I heard @kaspar030 is working on a DNS port based on sock. Is that correct?

Yes, I have a working synchronous dns client in https://github.com/kaspar030/sock/tree/master/dns, just need to find time to update nanocoap's makefile to include this.

**Comment by kaspar030** [comment]
This PR does a little more, but I don't have time to review...

**Comment by miri64** [comment]
Would review when sock-port is there :-)

**Comment by miri64** [comment]
@DipSwitch ping?

**Comment by miri64** [comment]
Postponed due to feature freeze

**Comment by emmanuelsearch** [comment]
@DipSwitch any news on the sock adaptation?

**Comment by miri64** [comment]
With #6694 next to being merged, I would like to close this as memo (since there are features in here, that are not implemented in #6694). Anyone disagreeing?

**Comment by PeterKietzmann** [comment]
I don't disagree. It seems both PRs are duplicates and unfortunately @DipSwitch hasn't bee available for quite some time. Reopen if I'm mistaken!

**Comment by PeterKietzmann** [comment]
Sorry @DipSwitch and thanks for the efforts you put into this. Hope to see you active again soon. Maybe you can provide help within #6694?!
